testnet/tools/nns-tools/lib/proposals.sh

```diff
@@ -83,40 +87,83 @@ generate_nns_upgrade_proposal_text() {
     fi
 
     OUTPUT=$(
-        cat <<EOF
-## Proposal to Upgrade the $CAPITALIZED_CANISTER_NAME Canister
-### Proposer: DFINITY Foundation
-### Git Hash: $NEXT_COMMIT
-### New Wasm Hash: $WASM_SHA$([ ! -z "$ARGS_HASH" ] && echo $'\n'"### Upgrade Args Hash: $ARGS_HASH")
-### Target canister: $(nns_canister_id "$CANISTER_NAME")
----
-## Features
-TODO ADD FEATURE NOTES
-## Release Notes
+        cat <<++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++
+# Upgrade the $CAPITALIZED_CANISTER_NAME Canister to Commit $SHORT_NEXT_COMMIT
+
+__Proposer__: ${PROPOSER}\\
+__Source Code__: [$NEXT_COMMIT][new-commit]
+
+[new-commit]: https://github.com/dfinity/ic/tree/$NEXT_COMMIT
+
+
+## Features, Fixes, and Optimizations
+
+TODO TO BE FILLED OUT BY THE PROPOSER
+
+
+## New Commits
+
 \`\`\`
 \$ git log --format="%C(auto) %h %s" $LAST_COMMIT..$NEXT_COMMIT --  $RELATIVE_CODE_LOCATION
 $(git log --format="%C(auto) %h %s" "$LAST_COMMIT".."$NEXT_COMMIT" -- $CANISTER_CODE_LOCATION)
-\`\`\`
-$([ ! -z "$CANDID_ARGS" ] && echo "## Candid Post Upgrade Args
+\`\`\`$(if [ ! -z "$CANDID_ARGS" ]; then
+            echo "
+
+
+## Upgrade Arguments
+
 \`\`\`candid
 $CANDID_ARGS
 \`\`\`
-### Validating Candid Args
-Verify that the encoded version of the plaintext args matches the \`arg_hex\` field in the proposal.
-\`didc encode '$CANDID_ARGS'\`
-")
-## Wasm Verification
-Verify that the hash of the gzipped WASM matches the proposed hash.
+"
+        fi)
+
+
+## Current Version
+
+- Current Git Hash: $LAST_COMMIT
+- Current Wasm Hash: $LAST_WASM_HASH
+
+
+## WASM Verification
+
+See ["Building the code"][prereqs] for prerequisites.
+
+[prereqs]: https://github.com/dfinity/ic?tab=readme-ov-file#building-the-code
+
 \`\`\`
+# 1. Get a copy of the code.
+git clone git@github.com:dfinity/ic.git
+cd ic
+# Or, if you already have a copy of the ic repo,
 git fetch
 git checkout $NEXT_COMMIT
+
+# 2. Build canisters.
 ./gitlab-ci/container/build-ic.sh -c
+
+# 3. Fingerprint the result.
 sha256sum ./artifacts/canisters/$(_canister_download_name_for_nns_canister_type "$CANISTER_NAME").wasm.gz
 \`\`\`
-## Current Version
-- Current Git Hash: $LAST_COMMIT
-- Current Wasm Hash: $LAST_WASM_HASH
-EOF
+
+This should match \`wasm_module\` field of this proposal.$(if [ ! -z "$CANDID_ARGS" ]; then
+            echo "
+
+
+## Upgrade Arguments Verification
+
+[\`didc\`][latest-didc] is required.
+
+[latest-didc]: https://github.com/dfinity/candid/releases/latest
+
+\`\`\`
+didc encode '$CANDID_ARGS' | sha256sum
```

Comment:
you don't actually sha256sum the candid args currently to verify.  Jason was working on updating this for some of the proposals, so that we get a sha256sum in the API response, at which point this will be correct.

But the sha256sum wouldn't match the arg_hex field.  I made this mistake in the past.

Reply:
Done.